crypto/src/vrf/transcript.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): S. Afach
+
+use merlin::Transcript;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TranscriptComponent {
+    RawData(Vec<u8>),
+    U64(u64),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct TranscriptAssembler {
+    label: &'static [u8],
+    components: Vec<(&'static [u8], TranscriptComponent)>,
+}
+
+// A wrapper that makes it unnecessary to directly use the merlin dependency
+#[derive(Clone)]
+pub struct WrappedTranscript(Transcript);
+
+impl From<Transcript> for WrappedTranscript {
+    fn from(t: Transcript) -> Self {
+        WrappedTranscript(t)
+    }
+}
+
+impl From<WrappedTranscript> for Transcript {
+    fn from(t: WrappedTranscript) -> Self {
+        t.0
+    }
+}
+
+impl TranscriptAssembler {
+    pub fn new(label: &'static [u8]) -> Self {
+        Self {
+            label,
+            components: Vec::new(),
+        }
+    }
+
+    pub fn attach(self, label: &'static [u8], value: TranscriptComponent) -> Self {
+        let mut result = self;
+        result.components.push((label, value));
+        result
```

Comment:
Yeah, I know... I don't like mut self though for my own (possibly stupid) biases... it doesn't make any sense. The value is being moved... why does it matter if it's gonna be changed in the interface? This is like in C++ doing a const string&&... that doesn't make sense. A moved value should always be considered mutable for a reader of an interface.

About the additional methods, I'd like to keep the touchable interface as small as possible to avoid having to test more. This is sensitive stuff.

Reply:
> Yeah, I know... I don't like mut self though for my own (possibly stupid) biases... it doesn't make any sense. The value is being moved... why does it matter if it's gonna be changed in the interface?

I see. I don't like that `mut` shows up in code like this. It is not, however, considered part of the interface. The tooling understand this: if you generate rustdoc or ask IDE for type of the method, it does not display the `mut` annotation.

Reply:
I really don't know which is worse... I have to think more about this. Though for what it's worth, I like that tooling understands the distinction... unlike in C++ where gmock wouldn't compile at times if you pass by const value in an interface.